processor/src/main/java/org/mapstruct/ap/internal/processor/creation/MappingResolverImpl.java

```diff
@@ -486,10 +511,9 @@ private Assignment resolveViaConversionAndMethod(Type sourceType, Type targetTyp
                 ySourceType = ySourceType.resolveTypeVarToType( targetType, methodYCandidate.getResultType() );
 
                 if ( ySourceType != null ) {
-                    methodRefY = resolveViaMethod( ySourceType, targetType, true );
+                    methodRefY = applyMatching( ySourceType, targetType, methodYCandidate );
                     if ( methodRefY != null ) {
-                        Type targetTypeX = methodYCandidate.getSourceParameters().get( 0 ).getType();
-                        ConversionAssignment conversionXRef = resolveViaConversion( sourceType, targetTypeX );
+                        ConversionAssignment conversionXRef = resolveViaConversion( sourceType, ySourceType );
```

Comment:
Why is this change needed?

`ySourceType` and `targetTypeX` don't have to be the same types I think. One can be more specific then the other.

Reply:
the change was forgotten last time.. it's actually a bug.

Reply:
👍